### ISSUE TITLE
[3.3.5a] "The Grand Melee" and "Among the Champions" not reliably giving credit

### ISSUE THREAD

**Issue by Treeston** [TrinityCore/TrinityCore#15534]
Even though pull #15015 was accepted, I'm still only seeing credit for these quests on about 20% of kills on revision c5cc7cca3e29e176b5ec0405a05c3db8e19c4e93.

This needs another look-over.

**Comment by sirikfoll** [comment]
it happens very rarely with me, my guess is that it's something with Line of Sight, console show a lot of map errors.
Maybe add a disable LoS for the spells?

**Comment by Treeston** [comment]
It's happening pretty consistently with me over the last few days. I did the quests on a Draenei Protection-spec paladin in case it matters.

**Comment by Killyana** [comment]
Before @sirikfoll fix the spell can fail and be casted on the mount instead of the player (even if the target is stored) actually if it fails the spell is not casted at all, so the problem is mounted players in movement cannot be targeted by the spell, it looks like a core issue.

**Comment by ccrs** [comment]
Try this

````diff
            void HandleScript(SpellEffIndex /*effIndex*/)
            {
-                 Unit* target = GetHitUnit()->GetCharmerOrOwnerOrSelf();
-                 target->CastSpell(target, uint32(GetEffectValue()));
+                if (Player* target = GetHitUnit()->GetCharmerOrOwnerPlayerOrPlayerItself())
+                     target->CastSpell(target, uint32(GetEffectValue()));
            }
````

**Comment by Killyana** [comment]
With your changes, the spell fails only if the last attack on the npc is a charge.

**Comment by Demonid** [comment]
Confirmed, @ccrs Your fix exactly has @Killyana mentioned.

**Comment by dr-j** [comment]
The only reports I have received about this is for some unknown reason credit not been given by Gnome valiants/champions and this is whats preventing players from obtaining lance a lot.

Cannot find any difference other than credit spells used by the gnomeregan npcs.

**Comment by Treeston** [comment]
Unless there was a change recently, credit is unreliable if the killing blow is from a charge. Killing blow works fine if you use thrust.

**Comment by Treeston** [comment]
fixed by #17870
